File: rest/src/main/java/my/sample/rest/impl/ItemRepository.java
package my.sample.rest.impl;

import my.sample.model.api.Item;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.List;

/**
 * @author mohammad shamsi <m.h.shams@gmail.com>
 */

public class ItemRepository {
    @PersistenceContext(unitName = "sampleDB")
    private EntityManager em;

    public void setEm(EntityManager em) {
        this.em = em;
    }

    public List<Item> get() {
        return em.createQuery("select i from Item as i", Item.class).getResultList();
    }

    public Item get(String id) {
        Item item = em.find(Item.class, id);
        if (item != null) {
            return item;
        }
        throw new IllegalArgumentException("Item not found. (item #" + id + ")");
    }

    public Item update(Item item) {
        Item i = em.find(Item.class, item.getId());
        if (i != null) {
            return em.merge(item);
        }
        throw new IllegalArgumentException("Item not found. (item #" + item.getId() + ")");
    }

    public Item add(Item item) {
        if (em.find(Item.class, item.getId()) != null) {
            throw new IllegalArgumentException("Duplicate item id. (item #" + item.getId() + ")");
        }
        em.persist(item);
        return item;
    }

    public Item delete(String id) {
        Item item = em.find(Item.class, id);
        if (item != null) {
            em.remove(item);
            return item;
        }
        throw new IllegalArgumentException("Item not found. (item #" + id + ")");
    }
}

File: rest/src/main/kotlin/my/sample/rest/impl/ItemRepository.kt
package my.sample.rest.impl

import my.sample.rest.api.Item

import java.util.ArrayList
import java.util.HashMap

/**
 * @author mohammad shamsi <m.h.shams@gmail.com>
 */

class ItemRepository {
    val items = hashMapOf<String, Item>();

    {
        for (i in 1..4) {
            val id = "$i"
            items.put(id,  Item.newItem(id, "Item #$id", "Description for item #$id"))
        }
    }

    fun get() = ArrayList(items.values())

    fun get(id: String): Item {
        if (items.containsKey(id)) {
            return items.get(id)!!
        }
        throw IllegalArgumentException("Item not found. (item #$id)")
    }

    fun update(item: Item): Item {
        if (items.containsKey(item.id)) {
            items.put(item.id!!, item)
            return items.get(item.id)!!
        }
        throw IllegalArgumentException("Item not found. (item #${item.id})")
    }

    fun add(item: Item): Item {
        if (items.containsKey(item.id)) {
            throw IllegalArgumentException("Duplicate item id. (item #${item.id})")
        }
        items.put(item.id!!, item)
        return items.get(item.id)!!
    }

    fun delete(id: String): Item {
        if (items.containsKey(id)) {
            return items.remove(id)!!
        }
        throw IllegalArgumentException("Item not found. (item #$id)")
    }
}
